### ISSUE TITLE
bug: Whisper "Segmentation fault (core dumped)" / inference got stuck

### ISSUE THREAD

**Issue by philschmid** [ELS-RD/kernl#272]
### Description

I tried to optimize Whipser using `Kernl` and the infernece gets "stuck"/Segmentation fault (core dumped). I am running on AWS on a A10G GPU. (g5 instance)

### Steps to reproduce

<details>
 <summary>1. Create Docker container </summary>

```bash
FROM nvidia/cuda:11.7.0-devel-ubuntu20.04
ENV DEBIAN_FRONTEND=noninteractive

ENV CUDA_INSTALL_PATH=/usr/local/cuda/
ENV CUDACXX=${CUDA_INSTALL_PATH}/bin/nvcc

WORKDIR /build
RUN apt-get update -y && \
 apt-get install -y software-properties-common ffmpeg && \
 add-apt-repository ppa:deadsnakes/ppa && \
 apt-get update -y

RUN apt-get install -y git \
    build-essential \
    libssl-dev \
    wget \
    curl \
    python3.9 \
    python3.9-distutils \
    python3.9-venv \
    python3.9-dev

RUN update-alternatives --install /usr/bin/python python /usr/bin/python3.8 1 && \
  update-alternatives --install /usr/bin/python python /usr/bin/python3.9 2 && \
  update-alternatives --set python /usr/bin/python3.9 && \
  update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 1 && \
  update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 2 && \
  update-alternatives --set python3 /usr/bin/python3.9

RUN python3.9 -m ensurepip --default-pip --upgrade

RUN pip install 'git+https://github.com/ELS-RD/kernl.git@v0.2.0' --extra-index-url https://download.pytorch.org/whl/nightly/cu117

# docker build -t kernl -f Dockerfile .
# docker run -it --rm --gpus all --workdir /home/ubuntu -v $(pwd)/:/home/ubuntu kernl /bin/bash
```
</details>

2. Start container `docker run -it --rm --gpus all --workdir /home/ubuntu -v $(pwd)/:/home/ubuntu kernl /bin/bash`

<details>
 <summary>3. Create `test_asr.py` script </summary>

```bash
from transformers import pipeline
from kernl.model_optimization import optimize_model
import torch 

# Load Transformer model and processor
asr = pipeline("automatic-speech-recognition", model="openai/whisper-medium", device=0,torch_dtype=torch.float16)

# optimize Encoder & Decoder with Kernl (in-place optimizations)
optimize_model(asr.model.model.encoder)
optimize_model(asr.model.model.decoder)

print(asr("1min_audio_sample.mp3"))
```
</details>

4. run script `python test_asr.py`

### Expected Behavior

Optimized inference with whisper

### Actual Behavior

Output get stuck 
```bash
/usr/local/lib/python3.9/dist-packages/transformers/generation/utils.py:1273: UserWarning: Neither `max_length` nor `max_new_tokens` has been set, `max_length` will default to 448 (`generation_config.max_length`). Controlling `max_length` via the config is deprecated and `max_length` will be removed from the config in v5 of Transformers -- we recommend using `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(
```

### Your environment

See Container file

### Self-service

- [X] I would be willing to help fix this bug myself.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by pommedeterresautee** [comment]
Tks a lot @philschmid for your report.
It's interesting, we have not made any test on ASR pipeline (not our use case).

You probably do not want to apply optimizations to the `encoder`, it matters very little in e2e times latency, increases the number of graphs to capture and CUDA graphs memory footprint.

I made your script work on a 2mn30 MP3 by setting cache_size_limit to 512 (I choose some moderately large random value) and removing the encoder related line.
However, it took a **very** long time to warmup (probably > 10mn, didn't measure).

Is there a doc describing how pipeline works? And what it brings vs not using it?

I am under the impression pipeline makes decoder work with longer and longer chain of tokens.
If true, it would make CUDA graphs unusable which would be sad as it provides most of a 2x speedup when Kernl is applied to whisper (even on large model!)

Please note that soon a PR for generative model inference should land in PyTorch 2.0.

**Comment by philschmid** [comment]
Thanks for checking in Michaël! 
> However, it took a very long time to warmup (probably > 10mn, didn't measure). 

Interesting! i think i canceled it < 10min waiting. Let me try again. 

> Is there a doc describing how pipeline works? And what it brings vs not using it?
> 
> I am under the impression pipeline makes decoder work with longer and longer chain of tokens.
> If true, it would make CUDA graphs unusable which would be sad as it provides most of a 2x speedup when Kernl is applied to whisper (even on large model!)

Your assumption is correct, the pipeline implements chunking and support for > 30s. Here is a blog: https://huggingface.co/blog/asr-chunking and here the documentation: https://huggingface.co/docs/transformers/v4.26.0/en/main_classes/pipelines#transformers.AutomaticSpeechRecognitionPipeline

> Please note that soon a PR for generative model inference should land in PyTorch 2.0.

Any thing you can share here? PR? or docs?

**Comment by pommedeterresautee** [comment]
**Before anything, disclaimer: I am not in anyway knowledgeable in ASR.**

That said, it's my understanding that Whisper has been trained on 30s audio chunks (https://arxiv.org/pdf/2212.04356.pdf § 2.3), and that it is part of the architecture of the model (see encoder tensor outputs, in particular shapes of K/V).

Therefore, during inference, it's kind of bad idea to not chunk if the file is > 30s long (out of distribution), which means, in the case of `Whisper`, default value of the pipeline for `chunk_length_s` parameter should not be `0`.
On the other side, the doc says that chunk_length_s is "only available for CTC models", which in my understanding is not the case of Whisper.

Is there a working chunking strategy for Whisper?
I am looking for something able to perform inference on - only - 30 seconds, and then aligning stride of the previous sequence with the current one.

**Comment by pommedeterresautee** [comment]
mmmhhh not sure what I did is right but this seems to work:

```python
from time import time

from transformers import pipeline, AutomaticSpeechRecognitionPipeline
from kernl.model_optimization import optimize_model
import torch
import torch._dynamo as dynamo

dynamo.config.cache_size_limit = 128
# Load Transformer model and processor
asr = pipeline("automatic-speech-recognition", model="openai/whisper-medium",
               device=0, torch_dtype=torch.float16, chunk_length_s=20, max_new_tokens=50)
# optimize Encoder & Decoder with Kernl (in-place optimizations)
# optimize_model(asr.model.model.encoder)
optimize_model(asr.model.model.decoder)

start = time()
print(asr("audio.mp3"))
print("1", time()-start)
start = time()
print(asr("audio.mp3"))
print("2", time()-start)
```

With Kernl: warmup 10mn, and 6s inference

```log
/home/geantvert/.local/share/virtualenvs/kernl/bin/python /home/geantvert/workspace/kernl/something.py 
1 581.9911861419678
{'text': " Allo ? Oui c'est papy là bonjour Vous êtes tombé ? Oh lala excusez-moi monsieur je voulais téléphoner à mon fils Je suis une personne âgée J'ai 85 ans monsieur Vous. Vous m'entendez? Oui, oui, très bien. 85 ans, hein? Oui, c'est bien. Eh oui, mais j' vois, je suis au bout du rouleau et là je sais pas comment faire. Bon je vous avoue, c'est pas pareil bien entendu mais il y a des fois vous savez il vaut mieux parler à quelqu'un qu qui vous arrive ? et bien c'est à dire que je... c'est embêtant àaperçois que je suis au bout du rouleau. Le papier ? Voilà. Si vous êtes pas loin de chez moi, est-ce que vous pourriez m'apporter un rouleau... un rouleau de papier hygiénique ? Ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха- ça porte bonheur vous voulez pas m'apporter un rouleau ? eh oui quand on est au bout du rouleau on comprend... Je vais quand même téléphoner à mon fils. Ouais. Oui parce que j'ai que le journal sur la main. Et ça gratte surtout quand on a des hémorroïdes. Merci ! Au revoir monsieur ! I'm sorry."}
2 6.391553163528442
```

without Kernl, 12s similar outptu, seems exepected:

```log
1 13.600375175476074
{'text': " Allo ? Oui c'est papy là bonjour Vous êtes tombé ? Oh lala excusez-moi monsieur je voulais téléphoner à mon fils Je suis une personne âgée J'ai 85 ans monsieur Vous. Vous m'entendez? Oui, oui, très bien. 85 ans, hein? Oui, c'est bien. Eh oui, mais j' vois, je suis au bout du rouleau et là je sais pas comment faire. Bon je vous avoue, c'est pas pareil bien entendu mais il y a des fois vous savez il vaut mieux parler à quelqu'un qu qui vous arrive ? Eh bien c'est à dire que je... C'est embêtant àaperçois que je suis au bout du rouleau. Le papier ? Voilà. Si vous êtes pas loin de chez moi, est-ce que vous pourriez m'apporter un rouleau... un rouleau de papier hygiénique ? Ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха-ха- ça porte bonheur vous voulez pas m'apporter un rouleau ? eh oui quand on est au bout du rouleau on comprend Je vais quand même téléphoner à mon fils Ouais Oui parce que j'ai que le journal sur la main. Et ça gratte surtout quand on a des hémorroïdes. Merci ! Au revoir monsieur ! I'm sorry."}
2 12.529566764831543
```

It's not very scientific measure, local run and I was doing other things on the machine, take it as simply indicative

**Comment by philschmid** [comment]
@pommedeterresautee i can confirm it works for me as well on a A10G, though the first inference took 24 minutes (1462s). 
Is there any way to accelerate the warm up? or run this a head of time and save/load it?

**Comment by pommedeterresautee** [comment]
wow 24mn! It's a bit curious, most of the warmup is CPU related and my machine is not that fast.
Warmup can be reduced but not removed, at least CUDA graph have to be done for each launch.

The Triton autotune part (15% of e2e times) can be saved/reused, PyTorch has some code in that spirit we can take inspiration from.
In the CUDA graphs capture part, we probably do an inference which is not required most of the time and I think it can be removed.

We plan to work on both things.

Other parts I can think of:
- layer replacement (Fx stuff), we made it 5x faster with this Kernl release (it was the slowest part of the warmup with a very bad scalability), we can have more cache here and there but would make the logic even more complicated, and it's already not super easy to follow. @gaetansnl what do you think?
- PyTorch side: Fx graph capture represents 40-50% of the e2e warmup time, done by dynamo, we do not touch this part. Truth is that there is no reason it can't be cached on disk, but not really seriously thought about it. May be we could ask to PyTorch team what they think about it?
